### ISSUE TITLE
Two rotary encoder

### ISSUE THREAD

**Issue by cypeks** [Edzelf/ESP32-Radio#63]
Is it possible to add a second rotary encoder? As in the traditional radio: one for volume control, the other for radio change, etc ...

**Comment by Edzelf** [comment]
Yes, you need 2 free input pins.  And you need to change the software.
It's easier to use a potentiometer instead of a rotary switch for the volume.

**Comment by cypeks** [comment]
Can you tell me how to adjust the volume in the program? Using the vs1053player.setVolume() object?

**Comment by Edzelf** [comment]
Just change ini_block.reqvol to a value between 0 and 100.

**Comment by Edzelf** [comment]
I just realized that you may connect 2 second rotary switch for just the volume in parallel of the first.  If you do not connect the SW-pin, the function of this switch will be limited to "volume" only.
No need to change the software and no extra I/O pins required.

**Comment by cypeks** [comment]
Not exactly good because I need to connect two rotary encoders. One to adjust the volume, the other to change the station.
I'll try to add enc2 support.
Where in the code are detected connected accessories? - I can not find it.

**Comment by Edzelf** [comment]
Yes, that's what I meant.  One for volume, one for the rest.

**Comment by DaniilMir** [comment]
Hello All,
I started implementation of two encoders.
Simply cloned ISR Routine and chk_enc routine. Also cloned all encoder variables.
But there is an issue with preset switching. Original routine has a hard-coded timeout : 
"case PRESET:
	{	waittime = 10;"
because of this timeout preset change works really slow comparing to the web interface command.
without timeout it works pretty fast, but I get another bug - encoder can skip presets.
It would be great to have a flag which shows that preset is already in switch process.
i tried : "if (datamode == STOPREQD) return; "
but it does not help. 

Even web interface command can switch two presets down without waiting for connecting to stream.
how can I identify that connecting process is already finished to allow second preset change command?

**Comment by Edzelf** [comment]
Yes, the rotary switch function is not very well implemented.  I will redesign it.

**Comment by koskee** [comment]
Here is a pretty good article I came across today in my inbox.. It shows only three shares, so I'm guessing it's new(ish). Although I haven't tried it, you might try this method out because I have found my encoder to be quite problematic (when using it, it will do any of the following: skip multiple increments, move in the wrong direction, or sometimes cause a crash/reboot).

This method sounds good on paper, from the description (the second method described) but as I haven't experimented with it, I do not know for sure. You might want to give it a try, since it would be as simple as copy/paste and could save you a bunch of time if it works well. 

 http://www.best-microcontroller-projects.com/rotary-encoder.html

**Comment by Edzelf** [comment]
That's exactly the method being used in the radio.  The difficulty is how to handle the results of the movements.

**Comment by DaniilMir** [comment]
best option is to split all functionality 
first: split encoder rotation and encoder clicks
second: spit encoder rotation and commands processing.

in current case we have scope to change volume and preset.
We have commands like  
"upvolume" -  "downvolume"  
"uppreset" - "downpreset"

on rotate action we can send this command and wait for a flag say"command process on"
till flag is on - ignore all other rotations. 
from encoder rotation process we expect direction only (even in current implementation)
flag processing can work even with existing setup with one encoder.  
we can bypass this flag with define option.

**Comment by koskee** [comment]
Well yes, it is similar in its use of tables, however I don't see any sort of denounce filter as he has described. 

I think that many of the issues (that I'm running into, anyway) could logically be attributed to contact bounce.. Either way, I'll try it when I  have a moment.

**Comment by Edzelf** [comment]
No need for debounce if a state machine is used to detect rotation.  I've improved the interface in the new version.  Added some debug output to show the detected clicks and rotation.

**Comment by dl6dbh** [comment]
Hallo  DaniilMir ,

have you finish the  implementation of two encoders? 
i am interested for an test 

Meinolf

**Comment by greppp** [comment]
> I just realized that you may connect 2 second rotary switch for just the volume in parallel of the first. If you do not connect the SW-pin, the function of this switch will be limited to "volume" only.
> No need to change the software and no extra I/O pins required.

Hi @Edzelf Edzelf, Could you elaborate a bit more on this? Would like to implement it. Thanks!

**Comment by Edzelf** [comment]
Plus, ground and the two rotation signals of the sevond encoder in parallel of the first.  What can I say more?

**Comment by greppp** [comment]
@Edzelf Thanks for your answer. Let's call "**encoder A**" the one that doesn't have the SW-pin connected and "**encoder B**" the one that has the SW-pin connected.
If I turn encoder A, the volume goes up or down. I get it. If I turn encoder B (knowing the rotations signal are in parallel of the other encoder), how come it will change the stations and not adjust the volume?
Thanks for your time, this is really appreciated.

**Comment by Edzelf** [comment]
I just test the radio with 2 encoders.  I put them in full parallel mode (all 5 pins).  It worked as expected.  I could use both encoders to change the volume.

**Comment by mousemat86** [comment]
I'm looking add adding a second encoder – encoder A for presets only and encoder B for volume only. 

Could any one explain running them in parallel – it's not clear from the above comments, are the encoders wired to seperate pins? Where would I declare the second encoder in the config? @Edzelf @greppp

**Comment by Edzelf** [comment]
Why not use a potentiometer for the volume?
For an extra rotary encoder you need at least 2 free I/O pins.  And some changes in the software.

**Comment by greppp** [comment]
@mousemat86 I never finished the project, but if you use a potentiometer for volume control then the added bonuses are that:
- It become more responsive
- Smooth transition from low to high volume (no steps), which will give you a nicer feel.
- No need for extra pins
- Less software change

good luck!
